include/ccf/historical_queries_adapter.h

```diff
@@ -83,9 +83,148 @@ namespace ccf::historical
+    const TxIDExtractor& extractor = txid_from_header)
+  {
+    return [f, &state_cache, available, extractor](
+             endpoints::EndpointContext& args) {
+      // Extract the requested transaction ID
+      ccf::TxID target_tx_id;
+      {
+        const auto tx_id_opt = extractor(args);
+        if (tx_id_opt.has_value())
+        {
+          target_tx_id = tx_id_opt.value();
+        }
+        else
+        {
+          return;
+        }
+      }
+
+      // Check that the requested transaction ID is available
+      {
+        auto error_reason = fmt::format(
+          "Transaction {} is not available.", target_tx_id.to_str());
+        auto is_available =
+          available(target_tx_id.view, target_tx_id.seqno, error_reason);
+        switch (is_available)
+        {
+          case HistoricalTxStatus::Error:
+            args.rpc_ctx->set_error(
+              HTTP_STATUS_INTERNAL_SERVER_ERROR,
+              ccf::errors::TransactionPendingOrUnknown,
+              std::move(error_reason));
+            return;
+          case HistoricalTxStatus::PendingOrUnknown:
+            // Set header No-Cache
+            args.rpc_ctx->set_response_header(
+              http::headers::CACHE_CONTROL, "no-cache");
+            args.rpc_ctx->set_error(
+              HTTP_STATUS_NOT_FOUND,
+              ccf::errors::TransactionPendingOrUnknown,
+              std::move(error_reason));
+            return;
+          case HistoricalTxStatus::Invalid:
+            args.rpc_ctx->set_error(
+              HTTP_STATUS_NOT_FOUND,
+              ccf::errors::TransactionInvalid,
+              std::move(error_reason));
+            return;
+          case HistoricalTxStatus::Valid:;
+        }
+      }
+
+      // We need a handle to determine whether this request is the 'same' as a
+      // previous one. For simplicity we use target_tx_id.seqno. This means we
+      // keep a lot of state around for old requests! It should be cleaned up
+      // manually
+      const auto historic_request_handle = target_tx_id.seqno;
+
+      // Get a state at the target version from the cache, if it is present
+      auto historical_state =
+        state_cache.get_state_at(historic_request_handle, target_tx_id.seqno);
+      if (historical_state == nullptr)
+      {
+        args.rpc_ctx->set_response_status(HTTP_STATUS_ACCEPTED);
+        static constexpr size_t retry_after_seconds = 3;
+        args.rpc_ctx->set_response_header(
+          http::headers::RETRY_AFTER, retry_after_seconds);
+        args.rpc_ctx->set_response_header(
+          http::headers::CONTENT_TYPE, http::headervalues::contenttype::TEXT);
+        args.rpc_ctx->set_response_body(fmt::format(
+          "Historical transaction {} is not currently available.",
+          target_tx_id.to_str()));
+        return;
+      }
+
+      // Call the provided handler
+      f(args, historical_state);
+    };
+  }
+
   static ccf::endpoints::EndpointFunction adapter_v1(
```

Comment:
I think we'd like to remove this in 2.0, so we should deprecate it now if possible (`CCF_DEPRECATED("Will be removed in 2.0")` in the right place).

Reply:
Yes